File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/service/HeartbeatService.java
package de.dimedis.mobileentry.service;

import android.app.AlarmManager;
import android.app.IntentService;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.util.Date;

import de.dimedis.mobileentry.backend.BackendService;
import de.dimedis.mobileentry.receiver.HeartbeatWakefulReceiver;
import de.dimedis.mobileentry.util.AppContext;
import de.dimedis.mobileentry.util.DataBaseUtil;
import de.dimedis.mobileentry.util.PrefUtils;
import de.dimedis.mobileentry.util.UpdateUtil;

public class HeartbeatService extends IntentService {
    static final String TAG = "HeartbeatService";

    public static void scheduleService() {
        Context context = AppContext.get();
        long updateIntervalSec = PrefUtils.getHeartbeatIntervalIdle();

        updateIntervalSec *= 1000L;
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent updateIntent = new Intent(context, HeartbeatService.class);
        PendingIntent pendingUpdateIntent = PendingIntent.getService(context, 0, updateIntent, PendingIntent.FLAG_CANCEL_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        // schedule to run in few minutes from now
        Date curr = new Date(System.currentTimeMillis());
        Date runAt = new Date(System.currentTimeMillis() + updateIntervalSec);

        if (curr.after(runAt)) {
            runAt = new Date(runAt.getTime() + updateIntervalSec);
            if (curr.after(runAt)) runAt = new Date(runAt.getTime() + updateIntervalSec);
        }
        alarmManager.cancel(pendingUpdateIntent);
        alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP, curr.getTime() + 1000/*runAt.getTime()*/, updateIntervalSec, pendingUpdateIntent);

        Log.i(TAG, "Update notification service scheduled to run at " + runAt);
        context.sendBroadcast(new Intent(context, HeartbeatWakefulReceiver.class));
    }


    public static void cancelService() {
        Context context = AppContext.get();
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent updateIntent = new Intent(context, HeartbeatService.class);
        PendingIntent pendingUpdateIntent = PendingIntent.getService(context, 0, updateIntent, PendingIntent.FLAG_CANCEL_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        alarmManager.cancel(pendingUpdateIntent);
        Log.i(TAG, "Update notification service cancelled");
    }

    public HeartbeatService() {
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        Log.i(TAG, "intent:" + intent);
        BackendService.sendHeartbeat();
        DataBaseUtil.uploadCachedTickets();
        DataBaseUtil.uploadCachedOfflineSessions();
        UpdateUtil.updateSettings();
    }
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/service/HeartbeatService.kt
package de.dimedis.mobileentry.service

import android.content.Intent
import android.app.IntentService
import android.app.AlarmManager
import android.app.PendingIntent
import android.util.Log
import de.dimedis.mobileentry.backend.BackendServiceUtil
import de.dimedis.mobileentry.receiver.HeartbeatWakefulReceiver
import de.dimedis.mobileentry.util.AppContext
import de.dimedis.mobileentry.util.DataBaseUtil
import de.dimedis.mobileentry.util.PrefUtils
import de.dimedis.mobileentry.util.UpdateUtil.updateSettings
import java.util.*

class HeartbeatService : IntentService(TAG) {
    override fun onHandleIntent(intent: Intent?) {
        Log.i(TAG, "intent:$intent")
        BackendServiceUtil.sendHeartbeat()
        DataBaseUtil.uploadCachedTickets()
        DataBaseUtil.uploadCachedOfflineSessions()
        updateSettings()
    }

    companion object {
        const val TAG = "HeartbeatService"
        fun scheduleService() {
            val context = AppContext.get()
            var updateIntervalSec: Long = PrefUtils.getHeartbeatIntervalIdle().toLong()
            updateIntervalSec *= 1000L
            val alarmManager = context.getSystemService(ALARM_SERVICE) as AlarmManager
            val updateIntent = Intent(context, HeartbeatService::class.java)
            val pendingUpdateIntent = PendingIntent.getService(context, 0, updateIntent, PendingIntent.FLAG_CANCEL_CURRENT)

            // schedule to run in few minutes from now
            val curr = Date(System.currentTimeMillis())
            var runAt = Date(System.currentTimeMillis() + updateIntervalSec)
            if (curr.after(runAt)) {
                runAt = Date(runAt.time + updateIntervalSec)
                if (curr.after(runAt)) runAt = Date(runAt.time + updateIntervalSec)
            }
            alarmManager.cancel(pendingUpdateIntent)
            alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP, curr.time + 1000, updateIntervalSec, pendingUpdateIntent)
            Log.i(TAG, "Update notification service scheduled to run at $runAt")
            context.sendBroadcast(Intent(context, HeartbeatWakefulReceiver::class.java))
        }

        fun cancelService() {
            val context = AppContext.get()
            val alarmManager = context.getSystemService(ALARM_SERVICE) as AlarmManager
            val updateIntent = Intent(context, HeartbeatService::class.java)
            val pendingUpdateIntent = PendingIntent.getService(context, 0, updateIntent, PendingIntent.FLAG_CANCEL_CURRENT)
            alarmManager.cancel(pendingUpdateIntent)
            Log.i(TAG, "Update notification service cancelled")
        }
    }
}